# Evaluation of RAG Using Ragas

In the following notebook we'll explore how to evaluate RAG pipelines using a powerful open-source tool called "Ragas". This will give us tools to evaluate component-wise metrics, as well as end-to-end metrics about the performance of our RAG pipelines.

In the following notebook we'll complete the following tasks:

- 🤝 Breakout Room Part #1:
  1. Install required libraries
  2. Set Environment Variables
  3. Creating a simple RAG pipeline with [LangChain v0.2.0](https://python.langchain.com/v0.2/docs/versions/v0_2/)
  4. Synthetic Dataset Generation for Evaluation using the [Ragas](https://github.com/explodinggradients/ragas) framework.
  

- 🤝 Breakout Room Part #2:
  1. Evaluating our pipeline with Ragas
  3. Making Adjustments to our RAG Pipeline
  4. Evaluating our Adjusted pipeline against our baseline
  5. Testing OpenAI's Claim

The only way to get started is to get started - so let's grab our dependencies for the day!

> NOTE: Using this notebook as presented will occur a charge of ~$3USD from OpenAI usage. Most of this cost is produced by the Synthetic Data Generation step - if you want to reduce costs, please use the provided commented code to leverage `GPT-3.5-Turbo` as the `critic_llm`!

## Motivation

A claim, made by OpenAI, is that their `text-embedding-3-small` is better (generally) than their `text-embedding-ada-002` model.

Here's some passages from their [blog](https://openai.com/blog/new-embedding-models-and-api-updates) about the `text-embedding-3` release:

> `text-embedding-3-small` is our new highly efficient embedding model and provides a significant upgrade over its predecessor, the `text-embedding-ada-002` model...

> **Stronger performance.** Comparing `text-embedding-ada-002` to `text-embedding-3-small`, the average score on a commonly used benchmark for multi-language retrieval ([MIRACL](https://github.com/project-miracl/miracl)) has increased from 31.4% to 44.0%, while the average score on a commonly used benchmark for English tasks ([MTEB](https://github.com/embeddings-benchmark/mteb)) has increased from 61.0% to 62.3%.

Well, with a library like Ragas - we can put that claim to the test!

If what they claim is true - we should see an increase on related metrics by using the new embedding model!

# 🤝 Breakout Room Part #1

## Task 1: Installing Required Libraries

A reminder that one of the [key features](https://blog.langchain.dev/langchain-v0-1-0/) of LangChain v0.1.0 is the compartmentalization of the various LangChain ecosystem packages!

So let's begin grabbing all of our LangChain related packages!

In [ ]:
!pip install -U -q langchain langchain-openai langchain_core langchain-community langchainhub openai

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 974.6/974.6 kB 6.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.3/40.3 kB 1.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 321.8/321.8 kB 6.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 28.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 327.4/327.4 kB 13.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.1/127.1 kB 6.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 17.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 75.6/75.6 kB 3.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.2/49.2 kB 2.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.9/77.9 kB 6.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.3/58.3 kB 5.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 145.0/145.0 kB 1.8 MB/s eta 0:00:00


We'll also get the "star of the show" today, which is Ragas!

In [ ]:
!pip install -qU ragas

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.1/86.1 kB 2.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 547.8/547.8 kB 11.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 71.1/71.1 kB 8.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.8/40.8 MB 21.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 13.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.9/64.9 kB 8.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 22.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 15.7 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
cudf-cu12 24.4.1 requires pyarrow<15.0.0a0,>=14.0.1, but you have pyarrow 16.1.0 which is incompatible.
google-colab 1.0.0 requires requests==2.31.0, but you have requests 

We'll be leveraging [QDrant](https://qdrant.tech/) again as our LangChain `VectorStore`.

We'll also install `pymupdf` and its dependencies which will allow us to load PDFs using the `PyMuPDFLoader` in the `langchain-community` package!

In [ ]:
!pip install -qU qdrant-client pymupdf pandas

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 230.1/230.1 kB 4.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.5/3.5 MB 46.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.0/13.0 MB 81.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.3/2.3 MB 61.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 15.8/15.8 MB 49.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 309.2/309.2 kB 28.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.5/57.5 kB 7.1 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
cudf-cu12 24.4.1 requires pandas<2.2.2dev0,>=2.0, but you have pandas 2.2.2 which is incompatible.
cudf-cu12 24.4.1 requires protobuf<5,>=3.20, but you have protobuf 5.27.1 which is incompatible.
cudf-cu12 24.4.1 requires pyarrow<15.0.0a0,>=14.0.1, but you hav

## Task 2: Set Environment Variables

Let's set up our OpenAI API key so we can leverage their API later on.

In [ ]:
import os
import openai
from getpass import getpass

openai.api_key = getpass("Please provide your OpenAI Key: ")
os.environ["OPENAI_API_KEY"] = openai.api_key

Please provide your OpenAI Key: ··········


## Task 3: Creating a Simple RAG Pipeline with LangChain v0.2.0

Building on what we learned last week, we'll be leveraging LangChain v0.2.0 and LCEL to build a simple RAG pipeline that we can baseline with Ragas.

## Building our RAG pipeline

Let's review the basic steps of RAG again:

- Create an Index
- Use retrieval to obtain pieces of context from our Index that are similar to our query
- Use a LLM to generate responses based on the retrieved context

Let's get started by creating our index.

> NOTE: We're going to start leaning on the term "index" to refer to our `VectorStore`, `VectorDatabase`, etc. We can think of "index" as the catch-all term, whereas `VectorStore` and the like relate to the specific technologies used to create, store, and interact with the index.

### Creating an Index

You'll notice that the largest changes (outside of some import changes) are that our old favourite chains are back to being bundled in an easily usable abstraction.

We can still create custom chains using LCEL - but we can also be more confident that our pre-packaged chains are creating using LCEL under the hood.

#### Loading Data

Let's start by loading some data!

> NOTE: You'll notice that we're using a document loader from the community package of LangChain. This is part of the v0.2.0 changes that make the base (`langchain-core`) package remain lightweight while still providing access to some of the more powerful community integrations.

In [ ]:
from langchain_community.document_loaders import PyMuPDFLoader

loader = PyMuPDFLoader(
    "https://d1lamhf6l6yk6d.cloudfront.net/uploads/2021/08/The-pmarca-Blog-Archives.pdf",
)

documents = loader.load()

In [ ]:
documents[0].metadata

{'source': 'https://d1lamhf6l6yk6d.cloudfront.net/uploads/2021/08/The-pmarca-Blog-Archives.pdf',
 'file_path': 'https://d1lamhf6l6yk6d.cloudfront.net/uploads/2021/08/The-pmarca-Blog-Archives.pdf',
 'page': 0,
 'total_pages': 195,
 'format': 'PDF 1.3',
 'title': 'The Pmarca Blog Archives',
 'author': '',
 'subject': '',
 'keywords': '',
 'creator': '',
 'producer': 'Mac OS X 10.10 Quartz PDFContext',
 'creationDate': "D:20150110020418Z00'00'",
 'modDate': "D:20150110020418Z00'00'",
 'trapped': ''}

#### Transforming Data

Now that we've got our single document - let's split it into smaller pieces so we can more effectively leverage it with our retrieval chain!

We'll start with the classic: `RecursiveCharacterTextSplitter`.

In [ ]:
from langchain.text_splitter import RecursiveCharacterTextSplitter

text_splitter = RecursiveCharacterTextSplitter(
    chunk_size = 200,
    chunk_overlap = 50
)

documents = text_splitter.split_documents(documents)

Let's confirm we've split our document.

In [ ]:
len(documents)

1864

#### Loading OpenAI Embeddings Model

We'll need a process by which we can convert our text into vectors that allow us to compare to our query vector.

Let's use OpenAI's `text-embedding-ada-002` for this task!

In [ ]:
from langchain_openai import OpenAIEmbeddings

embeddings = OpenAIEmbeddings(
    model="text-embedding-ada-002"
)

#### Creating a QDrant VectorStore

Now that we have documents - we'll need a place to store them alongside their embeddings.

In [ ]:
from langchain_community.vectorstores import Qdrant

qdrant_vector_store = Qdrant.from_documents(
    documents,
    embeddings,
    location=":memory:",
    collection_name="PMarca Blogs",
)

####❓ Question #1:

List out a few of the techniques that Qdrant uses that make it performant.

> NOTE: Check the [documentation](https://qdrant.tech/documentation/overview/) for more information about QDrant!

**ANSWER:**

**1/ Hierarchical Navigable Small World (HNSW) for specialized data structures and indexing techniques**

**2/ Product Quantization for more efficient memory and compute processes**

**3/ Support for various distance metrics e.g. Euclidean Distance, Cosine Similarity, and Dot Product to enable fast similarity and semantic search**

#### Creating a Retriever

To complete our index, all that's left to do is expose our vectorstore as a retriever - which we can do the same way we would in previous version of LangChain!

In [ ]:
retriever = qdrant_vector_store.as_retriever()

#### Testing our Retriever

Now that we've gone through the trouble of creating our retriever - let's see it in action!

In [ ]:
retrieved_documents = retriever.invoke("What is a rule of thumb for selecting an industry to invest in?")

In [ ]:
for doc in retrieved_documents:
  print(doc)

page_content='the existing order — and make sure that those forces of change\nhave a reasonable chance at succeeding.\nSecond rule of thumb:\nOnce you have picked an industry, get right to the center of it' metadata={'source': 'https://d1lamhf6l6yk6d.cloudfront.net/uploads/2021/08/The-pmarca-Blog-Archives.pdf', 'file_path': 'https://d1lamhf6l6yk6d.cloudfront.net/uploads/2021/08/The-pmarca-Blog-Archives.pdf', 'page': 125, 'total_pages': 195, 'format': 'PDF 1.3', 'title': 'The Pmarca Blog Archives', 'author': '', 'subject': '', 'keywords': '', 'creator': '', 'producer': 'Mac OS X 10.10 Quartz PDFContext', 'creationDate': "D:20150110020418Z00'00'", 'modDate': "D:20150110020418Z00'00'", 'trapped': '', '_id': '5029f7d838774eb99b1819cbeb059374', '_collection_name': 'PMarca Blogs'}
page_content='Third rule:\nIn a rapidly changing Held like technology, the best place to\nget experience when you’re starting out is in younger, high-\ngrowth companies.' metadata={'source': 'https://d1lamhf6l6yk6d

### Creating a RAG Chain

Now that we have the "R" in RAG taken care of - let's look at creating the "AG"!

#### Creating a Prompt Template

There are a few different ways we could create our prompt template - we could create a custom template, as seen in the code below, or we could simply pull a prompt from the prompt hub! Let's look at an example of that!

In [ ]:
from langchain import hub

retrieval_qa_prompt = hub.pull("langchain-ai/retrieval-qa-chat")

In [ ]:
print(retrieval_qa_prompt.messages[0].prompt.template)

Answer any use questions based solely on the context below:

<context>
{context}
</context>


As you can see - the prompt template is simple (and has a small error) - so we'll create our own to be a bit more specific!

In [ ]:
from langchain.prompts import ChatPromptTemplate

template = """Answer the question based only on the following context. If you cannot answer the question with the context, please respond with 'I don't know':

Context:
{context}

Question:
{question}
"""

prompt = ChatPromptTemplate.from_template(template)

#### Setting Up our Basic QA Chain

Now we can instantiate our basic RAG chain!

We'll use LCEL directly just to see an example of it - but you could just as easily use an abstraction here to achieve the same goal!

We'll also ensure to pass-through our context - which is critical for RAGAS.

In [ ]:
from operator import itemgetter

from langchain_openai import ChatOpenAI
from langchain_core.output_parsers import StrOutputParser
from langchain_core.runnables import RunnablePassthrough

primary_qa_llm = ChatOpenAI(model_name="gpt-3.5-turbo", temperature=0)

retrieval_augmented_qa_chain = (
    # INVOKE CHAIN WITH: {"question" : "<<SOME USER QUESTION>>"}
    # "question" : populated by getting the value of the "question" key
    # "context"  : populated by getting the value of the "question" key and chaining it into the base_retriever
    {"context": itemgetter("question") | retriever, "question": itemgetter("question")}
    # "context"  : is assigned to a RunnablePassthrough object (will not be called or considered in the next step)
    #              by getting the value of the "context" key from the previous step
    | RunnablePassthrough.assign(context=itemgetter("context"))
    # "response" : the "context" and "question" values are used to format our prompt object and then piped
    #              into the LLM and stored in a key called "response"
    # "context"  : populated by getting the value of the "context" key from the previous step
    | {"response": prompt | primary_qa_llm, "context": itemgetter("context")}
)

####🏗️ Activity #1:

Describe the pipeline shown above in simple terms. You can include a diagram if desired.

**ANSWER:**

**In the first runnable step {"context": itemgetter("question") | retriever, "question": itemgetter("question")}, the itemgetter("question") retrieves the user’s question and sends it down 2 parallel paths. The 2 output are "context" (i.e. the relevant information found by the retriever) and "question" (i.e the original question).**

**In the second runnable step | RunnablePassthrough.assign(context=itemgetter("context")), the itemgetter("context") extracts the context found in the first step and then passes it along to the next runnable step.**

**In the final runnable step {"response": prompt | primary_qa_llm, "context": itemgetter("context")}, the question and the retrieved context are combined into a format suitable for a language model (prompt). This meta-prompt is then sent to the LLM, which generates a response. Two outputs are produced -  "response" (the answer generated by the language model) and "context" (the original context is kept for reference or further use).**



Let's test it out!

In [ ]:
question = "What is a rule of thumb for selecting an industry to invest in?"

result = retrieval_augmented_qa_chain.invoke({"question" : question})

print(result["response"].content)

Get right to the center of it.


In [ ]:
question = "What did Pink Floyd have to say about how to proceed when investing in a new industry?"

result = retrieval_augmented_qa_chain.invoke({"question" : question})

print(result["response"].content)
print(result["context"])

I don't know.
[Document(page_content='ask if you can call them again if things change.\nTrust me — they’d much rather be saying “yes” than “no” —\nthey need all the good investments they can get.\nSecond, consider the environment.', metadata={'source': 'https://d1lamhf6l6yk6d.cloudfront.net/uploads/2021/08/The-pmarca-Blog-Archives.pdf', 'file_path': 'https://d1lamhf6l6yk6d.cloudfront.net/uploads/2021/08/The-pmarca-Blog-Archives.pdf', 'page': 15, 'total_pages': 195, 'format': 'PDF 1.3', 'title': 'The Pmarca Blog Archives', 'author': '', 'subject': '', 'keywords': '', 'creator': '', 'producer': 'Mac OS X 10.10 Quartz PDFContext', 'creationDate': "D:20150110020418Z00'00'", 'modDate': "D:20150110020418Z00'00'", 'trapped': '', '_id': 'dae3c45d18f84bef8853744bdf63320b', '_collection_name': 'PMarca Blogs'}), Document(page_content='watching carefully — if everyone agrees right up front that\nwhatever you are doing makes total sense, it probably isn’t a new\nand radical enough idea to justify a

We can already see that there are some improvements we could make here.

For now, let's switch gears to RAGAS to see how we can leverage that tool to provide us insight into how our pipeline is performing!

## Task 4: Synthetic Dataset Generation for Evaluation using Ragas

Ragas is a powerful library that lets us evaluate our RAG pipeline by collecting input/output/context triplets and obtaining metrics relating to a number of different aspects of our RAG pipeline.

We'll be evaluating on every core metric today, but in order to do that - we'll need to create a test set. Luckily for us, Ragas can do that directly!

### Synthetic Test Set Generation

We can leverage Ragas' [`Synthetic Test Data generation`](https://docs.ragas.io/en/stable/concepts/testset_generation.html) functionality to generate our own synthetic QC pairs - as well as a synthetic ground truth - quite easily!

In [ ]:
loader = PyMuPDFLoader(
    "https://d1lamhf6l6yk6d.cloudfront.net/uploads/2021/08/The-pmarca-Blog-Archives.pdf",
)

eval_documents = loader.load()

text_splitter_eval = RecursiveCharacterTextSplitter(
    chunk_size = 600,
    chunk_overlap = 50
)

eval_documents = text_splitter_eval.split_documents(eval_documents)

####❓ Question #2:

Why is it important to split our documents using different parameters when creating our synthetic data?

**ANSWER:**

**1/ Balanced data for training and eval: Splitting documents using different parameters helps in creating a balanced dataset where different types of context and content granularity are equally represented.**

**2/Realistic query generation: Splitting documents in different ways ensures that synthetic queries mimic real-world scenarios where users might seek information at different granularities.**

**3/Enhanced model generalization: When a model is trained or evaluated on data that has been split in various ways, it is exposed to a broader range of textual patterns and structures. This enhances the model's ability to generalize from the training data to unseen data, improving its performance in real-world applications.**


In [ ]:
len(eval_documents)

624


> NOTE: 🛑 Using this notebook as presented will occur a charge of ~$3USD from OpenAI usage. Most of this cost is produced by the Synthetic Data Generation step - if you want to reduce costs, please use the provided commented code to leverage GPT-3.5-Turbo as the critic_llm. If you're attempting to create a lot of samples please be aware of cost, as well as rate limits. 🛑

In [ ]:
from ragas.testset.generator import TestsetGenerator
from ragas.testset.evolutions import simple, reasoning, multi_context
from langchain_openai import ChatOpenAI, OpenAIEmbeddings

generator_llm = ChatOpenAI(model="gpt-3.5-turbo")
# critic_llm = ChatOpenAI(model="gpt-3.5-turbo") <--- If you don't have GPT-4 access, or to reduce cost/rate limiting issues.
critic_llm = ChatOpenAI(model="gpt-4o")
embeddings = OpenAIEmbeddings()

generator = TestsetGenerator.from_langchain(
    generator_llm,
    critic_llm,
    embeddings
)

distributions = {
    simple: 0.5,
    multi_context: 0.4,
    reasoning: 0.1
}

num_qa_pairs = 20 # You can reduce the number of QA pairs to 5 if you're experiencing rate-limiting issues

testset = generator.generate_with_langchain_docs(eval_documents[:200], num_qa_pairs, distributions)
testset.to_pandas()

embedding nodes:   0%|          | 0/400 [00:00<?, ?it/s]

Generating:   0%|          | 0/20 [00:00<?, ?it/s]

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



question  \
0   What is depicted in J. Ross Browne's Etchings ...   
1   How should a startup approach raising money be...   
2   How do the roles and responsibilities of emplo...   
3   What were some of the possible uses for the ph...   
4   How do great companies benefit from developing...   
5   How important is it to remain focused on produ...   
6   How can we make it easier for low-income peopl...   
7   How effective is submitting unsolicited busine...   
8   How does running out of cash impact the intens...   
9   How should a startup approach raising money be...   
10  What are the pros and cons of partnering with ...   
11  How can entrepreneurs support their startup pr...   
12  Can innovative products create new markets, li...   
13  How does ample funds impact a company's urgenc...   
14  How can a lax management culture impact a comp...   
15  How can startups benefit from corporate relati...   
16  Which company did Jim Clark head in Silicon Va...   
17  How does a startup team's decision-making proc...   

                                             contexts  \
0   [Part 5: The Moby Dick theory of big\ncompanie...   
1   [money a startup should raise and around which...   
2   [Second, in a startup, absolutely nothing happ...   
3   [description supplied by Edison’s assistant.\n...   
4   [Third, whenever possible, promote from within...   
5   [this cash! Creating a golden opportunity for ...   
6   [new way to work together, or anything else yo...   
7   [funds, she probably meets with 15 or 20, and ...   
8   [not so easy when you’re close to running out ...   
9   [money a startup should raise and around which...   
10  [less gulf of his paunch.”\n— Holland’s Plutar...   
11  [This obviously raises the issue of how you’re...   
12  [VMWare’s product was so profoundly transforma...   
13  [too much money is to hire too many people. Th...   
14  [you are successful, because you probably won’...   
15  [There are times in the life of a startup when...   
16  [The rest Yinched and didn’t do it.\nAnd this ...   
17  [(In case you were wondering, by the way, the ...   

                                         ground_truth evolution_type  \
0                                                 nan         simple   
1   Before Product/Market Fit, a startup should ai...         simple   
2   In a startup, employees have to make things ha...         simple   
3   Possible uses for the phonograph noted by Edis...         simple   
4   Great companies benefit from developing their ...         simple   
5   It is important to remain just as focused on p...         simple   
6   One way to make it easier for low-income peopl...         simple   
7   Submitting unsolicited business plans to a ven...         simple   
8   Running out of cash intensifies the challenges...         simple   
9   If a startup is before Product/Market Fit and ...         simple   
10  The downside of dealing with a big company is ...  multi_context   
11  Entrepreneurs can support their startup pre-fu...  multi_context   
12  Yes, innovative products like VMWare's operati...  multi_context   
13  Ample funds can lead to risks such as excessiv...  multi_context   
14  A lax management culture can have a significan...  multi_context   
15  Startups can benefit from corporate relationsh...  multi_context   
16  Jim Clark headed Silicon Graphics Inc. in Sili...      reasoning   
17  The decision-making process of a startup team ...      reasoning   

                                             metadata  episode_done  
0   [{'source': 'https://d1lamhf6l6yk6d.cloudfront...          True  
1   [{'source': 'https://d1lamhf6l6yk6d.cloudfront...          True  
2   [{'source': 'https://d1lamhf6l6yk6d.cloudfront...          True  
3   [{'source': 'https://d1lamhf6l6yk6d.cloudfront...          True  
4   [{'source': 'https://d1lamhf6l6yk6d.cloudfront...          True  
5   [{'source': 'https://d1lamhf6l6yk6d.cloudfront...          True  
6   [{'source': 'h

####❓ Question #3:

`{simple: 0.5, reasoning: 0.25, multi_context: 0.25}`

What exactly does this mapping refer to?

> NOTE: Check out the Ragas documentation on this generation process [here](https://docs.ragas.io/en/stable/concepts/testset_generation.html).

**ANSWER:**

**This means simple tests will be 50% of the generated cases, multi_context 25%, and reasoning 25%.**

**Reasoning means rewriting the question in a way that enhances the need for reasoning to answer it effectively.**

**Multi-Context means rewriting the question in a manner that necessitates information from multiple related sections or chunks to formulate an answer.**

Let's look at the output and see what we can learn about it!

In [ ]:
testset.test_data[1]

DataRow(question='How should a startup approach raising money before and after achieving Product/Market Fit?', contexts=['money a startup should raise and around which you should ori-\nent your thinking.\nIf you are Before Product/Market Fit and you can’t raise\nenough money in one shot to get to Product/Market Fit, then\nyou will need get as far as you can on each round and demon-\nstrate progress towards Product/Market Fit when you raise each\nnew round.\nIf you are ALer Product/Market Fit and you can’t raise enough\nmoney in one shot to fully exploit your opportunity, you have\na high-class problem and will probably — but not deXnitely —\nXnd that it gets continually easier to raise new money as you\nneed it.'], ground_truth="Before Product/Market Fit, a startup should aim to raise enough money to demonstrate progress towards Product/Market Fit with each new round. After achieving Product/Market Fit, if the startup can't raise enough money in one shot to fully exploit the opportunit

### Generating Responses with RAG Pipeline

Now that we have some QC pairs, and some ground truths, let's evaluate our RAG pipeline using Ragas.

The process is, again, quite straightforward - thanks to Ragas and LangChain!

Let's start by extracting our questions and ground truths from our create testset.

We can start by converting our test dataset into a Pandas DataFrame.

In [ ]:
test_df = testset.to_pandas()

In [ ]:
test_df

question  \
0   What is depicted in J. Ross Browne's Etchings ...   
1   How should a startup approach raising money be...   
2   How do the roles and responsibilities of emplo...   
3   What were some of the possible uses for the ph...   
4   How do great companies benefit from developing...   
5   How important is it to remain focused on produ...   
6   How can we make it easier for low-income peopl...   
7   How effective is submitting unsolicited busine...   
8   How does running out of cash impact the intens...   
9   How should a startup approach raising money be...   
10  What are the pros and cons of partnering with ...   
11  How can entrepreneurs support their startup pr...   
12  Can innovative products create new markets, li...   
13  How does ample funds impact a company's urgenc...   
14  How can a lax management culture impact a comp...   
15  How can startups benefit from corporate relati...   
16  Which company did Jim Clark head in Silicon Va...   
17  How does a startup team's decision-making proc...   

                                             contexts  \
0   [Part 5: The Moby Dick theory of big\ncompanie...   
1   [money a startup should raise and around which...   
2   [Second, in a startup, absolutely nothing happ...   
3   [description supplied by Edison’s assistant.\n...   
4   [Third, whenever possible, promote from within...   
5   [this cash! Creating a golden opportunity for ...   
6   [new way to work together, or anything else yo...   
7   [funds, she probably meets with 15 or 20, and ...   
8   [not so easy when you’re close to running out ...   
9   [money a startup should raise and around which...   
10  [less gulf of his paunch.”\n— Holland’s Plutar...   
11  [This obviously raises the issue of how you’re...   
12  [VMWare’s product was so profoundly transforma...   
13  [too much money is to hire too many people. Th...   
14  [you are successful, because you probably won’...   
15  [There are times in the life of a startup when...   
16  [The rest Yinched and didn’t do it.\nAnd this ...   
17  [(In case you were wondering, by the way, the ...   

                                         ground_truth evolution_type  \
0                                                 nan         simple   
1   Before Product/Market Fit, a startup should ai...         simple   
2   In a startup, employees have to make things ha...         simple   
3   Possible uses for the phonograph noted by Edis...         simple   
4   Great companies benefit from developing their ...         simple   
5   It is important to remain just as focused on p...         simple   
6   One way to make it easier for low-income peopl...         simple   
7   Submitting unsolicited business plans to a ven...         simple   
8   Running out of cash intensifies the challenges...         simple   
9   If a startup is before Product/Market Fit and ...         simple   
10  The downside of dealing with a big company is ...  multi_context   
11  Entrepreneurs can support their startup pre-fu...  multi_context   
12  Yes, innovative products like VMWare's operati...  multi_context   
13  Ample funds can lead to risks such as excessiv...  multi_context   
14  A lax management culture can have a significan...  multi_context   
15  Startups can benefit from corporate relationsh...  multi_context   
16  Jim Clark headed Silicon Graphics Inc. in Sili...      reasoning   
17  The decision-making process of a startup team ...      reasoning   

                                             metadata  episode_done  
0   [{'source': 'https://d1lamhf6l6yk6d.cloudfront...          True  
1   [{'source': 'https://d1lamhf6l6yk6d.cloudfront...          True  
2   [{'source': 'https://d1lamhf6l6yk6d.cloudfront...          True  
3   [{'source': 'https://d1lamhf6l6yk6d.cloudfront...          True  
4   [{'source': 'https://d1lamhf6l6yk6d.cloudfront...          True  
5   [{'source': 'https://d1lamhf6l6yk6d.cloudfront...          True  
6   [{'source': 'h

In [ ]:
test_questions = test_df["question"].values.tolist()
test_groundtruths = test_df["ground_truth"].values.tolist()

Now we'll generate responses using our RAG pipeline using the questions we've generated - we'll also need to collect our retrieved contexts for each question.

We'll do this in a simple loop to see exactly what's happening!

In [ ]:
answers = []
contexts = []

for question in test_questions:
  response = retrieval_augmented_qa_chain.invoke({"question" : question})
  answers.append(response["response"].content)
  contexts.append([context.page_content for context in response["context"]])

Now we can wrap our information in a Hugging Face dataset for use in the Ragas library.

In [ ]:
from datasets import Dataset

response_dataset = Dataset.from_dict({
    "question" : test_questions,
    "answer" : answers,
    "contexts" : contexts,
    "ground_truth" : test_groundtruths
})

Let's take a peek and see what that looks like!

In [ ]:
response_dataset[1]

{'question': 'How should a startup approach raising money before and after achieving Product/Market Fit?',
 'answer': 'Before achieving Product/Market Fit, a startup should ideally raise enough money to get to Product/Market Fit. After achieving Product/Market Fit, a startup should ideally raise enough money to fully exploit the opportunity in front of it.',
 'contexts': ['This obviously raises the issue of how you’re supposed to do\nthat before you’ve raised money.\nTry to raise angel money, or bootstrap oW of initial customers',
  'Market Fit, and AMer Product/Market Fit.\nBefore Product/Market Fit, a startup should ideally raise at\nleast enough money to get to Product/Market Fit.',
  'this cash! Creating a golden opportunity for a smaller,\nscrappier startup to come along and kick your rear.\nSo what should you do if you do raise a lot of money?',
  'least enough money to get to Product/Market Fit.\nALer Product/Market Fit, a startup should ideally raise at least\nenough money to f

# 🤝 Breakout Room Part #2

## Task 1: Evaluating our Pipeline with Ragas

Now that we have our response dataset - we can finally get into the "meat" of Ragas - evaluation!

First, we'll import the desired metrics, then we can use them to evaluate our created dataset!

Check out the specific metrics we'll be using in the Ragas documentation:

- [Faithfulness](https://docs.ragas.io/en/stable/concepts/metrics/faithfulness.html)
- [Answer Relevancy](https://docs.ragas.io/en/stable/concepts/metrics/answer_relevance.html)
- [Context Precision](https://docs.ragas.io/en/stable/concepts/metrics/context_precision.html)
- [Context Recall](https://docs.ragas.io/en/stable/concepts/metrics/context_recall.html)
- [Answer Correctness](https://docs.ragas.io/en/stable/concepts/metrics/answer_correctness.html)

See the accompanied presentation for more in-depth explanations about each of the metrics!

In [ ]:
from ragas import evaluate
from ragas.metrics import (
    faithfulness,
    answer_relevancy,
    answer_correctness,
    context_recall,
    context_precision,
)

metrics = [
    faithfulness,
    answer_relevancy,
    context_recall,
    context_precision,
    answer_correctness,
]

All that's left to do is call "evaluate" and away we go!

In [ ]:
results = evaluate(response_dataset, metrics, raise_exceptions=False)

In [ ]:
results

{'faithfulness': 0.6637, 'answer_relevancy': 0.6974, 'context_recall': 0.7021, 'context_precision': 0.7706, 'answer_correctness': 0.7175}

In [ ]:
results_df = results.to_pandas()
results_df

question  \
0   What is depicted in J. Ross Browne's Etchings ...   
1   How should a startup approach raising money be...   
2   How do the roles and responsibilities of emplo...   
3   What were some of the possible uses for the ph...   
4   How do great companies benefit from developing...   
5   How important is it to remain focused on produ...   
6   How can we make it easier for low-income peopl...   
7   How effective is submitting unsolicited busine...   
8   How does running out of cash impact the intens...   
9   How should a startup approach raising money be...   
10  What are the pros and cons of partnering with ...   
11  How can entrepreneurs support their startup pr...   
12  Can innovative products create new markets, li...   
13  How does ample funds impact a company's urgenc...   
14  How can a lax management culture impact a comp...   
15  How can startups benefit from corporate relati...   
16  Which company did Jim Clark head in Silicon Va...   
17  How does a startup team's decision-making proc...   

                                               answer  \
0                        Whales and a whaling cruise.   
1   Before achieving Product/Market Fit, a startup...   
2   In a startup, employees have to take on multip...   
3   Some of the possible uses for the phonograph n...   
4   Great companies benefit from developing their ...   
5   It is important to remain focused on product a...   
6                                      Start Prosper.   
7   Submitting unsolicited business plans to a ven...   
8   Running out of cash can increase the intensity...   
9   Before achieving Product/Market Fit, a startup...   
10  Pros: Potential for partnership or distributio...   
11  Entrepreneurs can support their startup pre-fu...   
12  Yes, innovative products like VMWare can creat...   
13  Ample funds can lead to a false sense of accom...   
14  A lax management culture can impact a company'...   
15  Startups can benefit from corporate relationsh...   
16  Jim Clark headed Netscape in Silicon Valley in...   
17  The decision-making process of a startup team ...   

                                             contexts  \
0   [— bo-o-os!”\n“How far oE?”\n“Two miles and a ...   
1   [This obviously raises the issue of how you’re...   
2   [executive skill sets required for a big compa...   
3   [a list of possible uses for the phonograph wa...   
4   [Third, whenever possible, promote from within...   
5   [This obviously raises the issue of how you’re...   
6   [for low-income people to borrow money? Start ...   
7   [unsolicited, to a venture Xrm is likely to am...   
8   [ing. Keeping moving fast if you want to survi...   
9   [This obviously raises the issue of how you’re...   
10  [There are times in the life of a startup when...   
11  [funding are, in my view: angel funding, boots...   
12  [VMWare’s product was so profoundly transforma...   
13  [tant than worrying too much about dilution or...   
14  [culture to get set where the manager’s job is...   
15  [There are times in the life of a startup when...   
16  [companies as Sun, Cisco, Yahoo, and Google, s...   
17  [answer, in part because in the beginning of a...   

                                         ground_truth  faithfulness  \
0                                                 nan           NaN   
1   Before Product/Market Fit, a startup should ai...      1.000000   
2   In a startup, employees have to make things ha...      1.000000   
3   Possible uses for the phonograph noted by Edis...      0.500000   
4   Great companies benefit from developing their ...      1.000000   
5   It is important to remain just as focused on p...      0.333333   
6   One way to make it easier for low-income peopl...      1.000000   
7   Submitting unsolicited business plans to a ven...      0.500000   
8   Running out of cash intensifies the challenges...      1.000000   
9   If a startup is before Product/Market Fit and ...      0.500000   
10  The down

## Task 2: Making Adjustments to our RAG Pipeline

Now that we have established a baseline - we can see how any changes impact our pipeline's performance!

Let's modify our retriever and see how that impacts our Ragas metrics!

> NOTE: MultiQueryRetriever is expanded on [here](https://python.langchain.com/docs/modules/data_connection/retrievers/MultiQueryRetriever) but for now, the implementation is not important to our lesson!

In [ ]:
from langchain.retrievers import MultiQueryRetriever

advanced_retriever = MultiQueryRetriever.from_llm(retriever=retriever, llm=primary_qa_llm)

We'll also re-create our RAG pipeline using the abstractions that come packaged with LangChain v0.1.0!

First, let's create a chain to "stuff" our documents into our context!

In [ ]:
from langchain.chains.combine_documents import create_stuff_documents_chain

document_chain = create_stuff_documents_chain(primary_qa_llm, retrieval_qa_prompt)

Next, we'll create the retrieval chain!

In [ ]:
from langchain.chains import create_retrieval_chain

retrieval_chain = create_retrieval_chain(advanced_retriever, document_chain)

In [ ]:
response = retrieval_chain.invoke({"input": "Who is Taylor Swift fueding with?"})

In [ ]:
print(response["answer"])

I'm sorry, but based on the provided context, I cannot determine who Taylor Swift is feuding with.


In [ ]:
response = retrieval_chain.invoke({"input": "Why are they fueding?"})

In [ ]:
print(response["answer"])

The text does not provide specific information about why individuals or groups may be feuding.


Well, just from those responses this chain *feels* better - but lets see how it performs on our eval!

Let's do the same process we did before to collect our pipeline's contexts and answers.

In [ ]:
answers = []
contexts = []

for question in test_questions:
  response = retrieval_chain.invoke({"input" : question})
  answers.append(response["answer"])
  contexts.append([context.page_content for context in response["context"]])

Now we can convert this into a dataset, just like we did before.

In [ ]:
response_dataset_advanced_retrieval = Dataset.from_dict({
    "question" : test_questions,
    "answer" : answers,
    "contexts" : contexts,
    "ground_truth" : test_groundtruths
})

Let's evaluate on the same metrics we did for the first pipeline and see how it does!

In [ ]:
advanced_retrieval_results = evaluate(response_dataset_advanced_retrieval, metrics)

Evaluating:   0%|          | 0/100 [00:00<?, ?it/s]

In [ ]:
advanced_retrieval_results_df = advanced_retrieval_results.to_pandas()
advanced_retrieval_results_df

question  \
0   What are the potential risks and challenges fa...   
1   How can a transformative deal inject energy ba...   
2   Who is quoted as saying "This fucking job ain'...   
3   What is Los Angeles known for in terms of oppo...   
4   How can blogging about their point of view hel...   
5   Are the anticipated career peaks for creators ...   
6   What is the impact of accelerated vesting on c...   
7   What is the benefit of combining a useful grad...   
8   What are some ways to bootstrap a business bef...   
9   What is Richard Morgan's opinion of Peter Hami...   
10  What are the main responsibilities of a startu...   
11  How does the productivity-creativity relations...   
12  What are the consequences and trade-offs of th...   
13  What is the correlation between startup team c...   
14  How do technical degrees demonstrate commitmen...   
15  According to "A Perfect Mess," what is one tac...   
16  What was MySpace's revenue in 2007, given the ...   
17  How can interruptions be minimized and focus m...   
18  Which folder is for follow-up on commitments m...   
19  How does perceived risk impact decision-making...   

                                               answer  \
0   Not raising enough money risks the survival of...   
1   A transformative deal can inject energy back i...   
2   The quote "This fucking job ain't that fucking...   
3   Los Angeles is known for opportunities and ind...   
4   Blogging about their point of view can help an...   
5   The anticipated career peaks for creators who ...   
6   Accelerated vesting on change of control can i...   
7   Combining a useful graduate degree with a subs...   
8   Some ways to bootstrap a business before raisi...   
9   Richard Morgan views Peter Hamilton as the cle...   
10  A startup executive must take personal respons...   
11  The relationship between productivity and crea...   
12  The consequences and trade-offs of the money r...   
13  The caliber of a startup team can greatly impa...   
14  Technical degrees demonstrate commitment to fu...   
15  One tactic that can increase productivity by a...   
16  MySpace's revenue in 2007 was nearly $300 mill...   
17  Interruptions can be minimized and focus maint...   
18  The folder for follow-up on commitments made b...   
19  Perceived risk can impact decision-making in a...   

                                             contexts  \
0   [Here’s why you shouldn’t do that:\nWhat are t...   
1   [ossiXed that the clear opportunity exists to ...   
2   [then doesn’t know what the friend is currentl...   
3   [the city where all the action is happening.\n...   
4   [esting things going on, about their point of ...   
5   [early, end late, and produce at above-average...   
6   [Give her a stock-option grant with accelerate...   
7   [degree, you are much better oW combining it w...   
8   [This obviously raises the issue of how you’re...   
9   [ders and executions around the clock. Haven’t...   
10  [more as administrators and bureaucrats.\n•\nB...   
11  [the connection between productive precocity, ...   
12  [Here’s why you shouldn’t do that:\nWhat are t...   
13  [Let’s start by deXning terms.\nThe caliber of...   
14  [path to doing anything big.\n•\nPlus, technic...   
15  [By not keeping a schedule, I mean: refuse to ...   
16  [enue in 2007, and more in 2008.\nAs an indepe...   
17  [one.\nHide in an iPod\nOne of the best and ea...   
18  [the normal course of your day.\nFourth, aside...   
19  [What are the layers of risk for a high-tech\n...   

                                         ground_truth  faithfulness  \
0   Not raising enough money risks the survival of...      0.666667   
1                                                 nan      0.500000   
2                             The great Tommy Lasorda      1.000000   
3   Los Angeles is known for opportunities and ind...      1.000000   
4   Blogging about their point of view can help an...      1.000000   
5                        

## Task 3: Evaluating our Adjusted Pipeline Against Our Baseline

Now we can compare our results and see what directional changes occured!

Let's refresh with our initial metrics.

In [ ]:
results

{'faithfulness': 0.6255, 'answer_relevancy': 0.8526, 'context_recall': 0.6500, 'context_precision': 0.7583, 'answer_correctness': 0.5520}

And see how our advanced retrieval modified our chain!

In [ ]:
advanced_retrieval_results

{'faithfulness': 0.7120, 'answer_relevancy': 0.9455, 'context_recall': 0.7417, 'context_precision': 0.7112, 'answer_correctness': 0.5529}

In [ ]:
import pandas as pd

df_original = pd.DataFrame(list(results.items()), columns=['Metric', 'Baseline'])
df_comparison = pd.DataFrame(list(advanced_retrieval_results.items()), columns=['Metric', 'MultiQueryRetriever with Document Stuffing'])

df_merged = pd.merge(df_original, df_comparison, on='Metric')

df_merged['Delta'] = df_merged['MultiQueryRetriever with Document Stuffing'] - df_merged['Baseline']

df_merged

Metric  Baseline  MultiQueryRetriever with Document Stuffing  \
0        faithfulness  0.625490                                    0.712024   
1    answer_relevancy  0.852605                                    0.945499   
2      context_recall  0.650000                                    0.741667   
3   context_precision  0.758333                                    0.711185   
4  answer_correctness  0.551982                                    0.552914   

      Delta  
0  0.086534  
1  0.092894  
2  0.091667  
3 -0.047149  
4  0.000932

## Task 4: Testing OpenAI's Claim

Now that we've seen how our retriever can impact the performance of our RAG pipeline - let's see how changing our embedding model impacts performance.

####🏗️ Activity #2:

Please provide markdown, or code comments, to explain which each of the following steps are doing!

In [ ]:
new_embeddings = OpenAIEmbeddings(model="text-embedding-3-small")

In [ ]:
vector_store = Qdrant.from_documents(
    documents,
    new_embeddings,
    location=":memory:",
    collection_name="PMarca Blogs - TE3 - MQR",
)

In [ ]:
new_retriever = vector_store.as_retriever()

In [ ]:
new_advanced_retriever = MultiQueryRetriever.from_llm(retriever=new_retriever, llm=primary_qa_llm)

In [ ]:
new_retrieval_chain = create_retrieval_chain(new_advanced_retriever, document_chain)

In [ ]:
answers = []
contexts = []

for question in test_questions:
  response = new_retrieval_chain.invoke({"input" : question})
  answers.append(response["answer"])
  contexts.append([context.page_content for context in response["context"]])

In [ ]:
new_response_dataset_advanced_retrieval = Dataset.from_dict({
    "question" : test_questions,
    "answer" : answers,
    "contexts" : contexts,
    "ground_truth" : test_groundtruths
})

In [ ]:
new_advanced_retrieval_results = evaluate(new_response_dataset_advanced_retrieval, metrics)

Evaluating:   0%|          | 0/100 [00:00<?, ?it/s]

In [ ]:
new_advanced_retrieval_results

{'faithfulness': 0.8658, 'answer_relevancy': 0.9454, 'context_recall': 0.6792, 'context_precision': 0.8472, 'answer_correctness': 0.5055}

In [ ]:
df_baseline = pd.DataFrame(list(results.items()), columns=['Metric', 'ADA + Baseline'])
df_original = pd.DataFrame(list(advanced_retrieval_results.items()), columns=['Metric', 'ADA + MQR'])
df_comparison = pd.DataFrame(list(new_advanced_retrieval_results.items()), columns=['Metric', 'TE3 + MQR'])

df_merged = pd.merge(df_original, df_comparison, on='Metric')
df_merged = pd.merge(df_baseline, df_merged, on="Metric")

df_merged['ADA + MQR -> TE3 + MQR'] = df_merged['TE3 + MQR'] - df_merged['ADA + MQR']
df_merged['Baseline -> TE3 + MQR'] = df_merged['TE3 + MQR'] - df_merged['ADA + Baseline']

df_merged

Metric  ADA + Baseline  ADA + MQR  TE3 + MQR  \
0        faithfulness        0.625490   0.712024   0.865833   
1    answer_relevancy        0.852605   0.945499   0.945435   
2      context_recall        0.650000   0.741667   0.679167   
3   context_precision        0.758333   0.711185   0.847174   
4  answer_correctness        0.551982   0.552914   0.505525   

   ADA + MQR -> TE3 + MQR  Baseline -> TE3 + MQR  
0                0.153810               0.240343  
1               -0.000064               0.092831  
2               -0.062500               0.029167  
3                0.135990               0.088841  
4               -0.047388              -0.046457

####❓ Question #4:

Do you think, in your opinion, `text-embedding-3-small` is significantly better than `ada`?

**SUMMARY ANSWER:**

Based on the difference analysis below, TE3 excels in certain metrics like faithfulness and context precision, but ADA remains competitive in terms of context recall and answer correctness.

**DIFFERENCE ANALYSIS:**

Column ADA + MQR -> TE3 + MQR: Indicates the change in performance when switching from ADA to TE3 with MQR.

- Positive values for faithfulness (0.153810) and context precision (0.135990) suggest TE3 improves over ADA in these areas.
- Negative values for context recall (-0.062500) and answer correctness (-0.047388) suggest ADA performs better in these areas.

Baseline -> TE3 + MQR: Shows the improvement over the baseline when using TE3 with MQR.
- Faithfulness (0.240343), answer relevancy (0.092831) and context precision (0.088841) are significantly improved with TE3 compared to the baseline.


## BONUS ACTIVITY: Using a Better Generator

Now that we've seen how much more effective a better Retrieval pipeline is, let's look at what impact a better(?) Generator is!

Adapt the above `TE3 + MQR` pipeline to use `GPT-4o` and compare the results below!

--------ANSWER----------

Let's start by defining a new RAG chain using GP-4o as the LLM model of choice.

In [ ]:
primary_qa_llm = ChatOpenAI(model_name="gpt-3.5-turbo", temperature=0) # from before
secondary_qa_llm = ChatOpenAI(model_name="gpt-4o", temperature=0)

second_retrieval_augmented_qa_chain = (
    # INVOKE CHAIN WITH: {"question" : "<<SOME USER QUESTION>>"}
    # "question" : populated by getting the value of the "question" key
    # "context"  : populated by getting the value of the "question" key and chaining it into the base_retriever
    {"context": itemgetter("question") | retriever, "question": itemgetter("question")}
    # "context"  : is assigned to a RunnablePassthrough object (will not be called or considered in the next step)
    #              by getting the value of the "context" key from the previous step
    | RunnablePassthrough.assign(context=itemgetter("context"))
    # "response" : the "context" and "question" values are used to format our prompt object and then piped
    #              into the LLM and stored in a key called "response"
    # "context"  : populated by getting the value of the "context" key from the previous step
    | {"response": prompt | secondary_qa_llm, "context": itemgetter("context")}
)

Let's do a simple test to compare it with GPT-3.5-Turbo.

In [ ]:
question = "How should a startup approach raising money before and after achieving Product Market Fit??"
result1 = retrieval_augmented_qa_chain.invoke({"question" : question})
result2 = second_retrieval_augmented_qa_chain.invoke({"question" : question})
print("GTP-3.5-Turbo: " + result1["response"].content)
print("----")
print("GTP-4o: " + result2["response"].content)

GTP-3.5-Turbo: Before achieving Product/Market Fit, a startup should ideally raise enough money to get to Product/Market Fit. After achieving Product/Market Fit, a startup should ideally raise enough money to fully exploit the opportunity in front of it.
----
GTP-4o: Before achieving Product/Market Fit, a startup should ideally raise at least enough money to get to Product/Market Fit. After achieving Product/Market Fit, a startup should ideally raise at least enough money to fully exploit the opportunity in front of it.


OK, let's adapt the above TE3 + MQR pipeline.

In [ ]:
new_embeddings = OpenAIEmbeddings(model="text-embedding-3-small")

In [ ]:
vector_store = Qdrant.from_documents(
    documents,
    new_embeddings,
    location=":memory:",
    collection_name="PMarca Blogs - TE3 - MQR",
)

In [ ]:
new_retriever = vector_store.as_retriever()

In [ ]:
# new_advanced_retriever = MultiQueryRetriever.from_llm(retriever=new_retriever, llm=primary_qa_llm)
second_new_advanced_retriever = MultiQueryRetriever.from_llm(retriever=new_retriever, llm=secondary_qa_llm)

In [ ]:
second_new_retrieval_chain = create_retrieval_chain(second_new_advanced_retriever, document_chain)

In [ ]:
answers2 = []
contexts2 = []

for question in test_questions:
  response = second_new_retrieval_chain.invoke({"input" : question})
  answers2.append(response["answer"])
  contexts2.append([context.page_content for context in response["context"]])

In [ ]:
second_new_response_dataset_advanced_retrieval = Dataset.from_dict({
    "question" : test_questions,
    "answer" : answers2,
    "contexts" : contexts2,
    "ground_truth" : test_groundtruths
})

In [ ]:
second_new_response_dataset_advanced_retrieval[1]

{'question': 'How should a startup approach raising money before and after achieving Product/Market Fit?',
 'answer': "Before achieving Product/Market Fit, a startup should ideally raise at least enough money to get to Product/Market Fit. This means securing funding to develop and refine the product until it resonates with the market. After achieving Product/Market Fit, a startup should ideally raise enough money to fully exploit the opportunity in front of it. This involves scaling the business, expanding market reach, and maximizing growth potential. It's important for startups to demonstrate progress towards Product/Market Fit when raising each new round of funding.",
 'contexts': ['priate for your particular startup.\nWith a working product that could be the foundation of a fund-\nable startup, you have a much better chance of getting funded',
  'ent your thinking.\nIf you are Before Product/Market Fit and you can’t raise\nenough money in one shot to get to Product/Market Fit, then

In [ ]:
second_new_advanced_retrieval_results = evaluate(second_new_response_dataset_advanced_retrieval, metrics)

Evaluating:   0%|          | 0/90 [00:00<?, ?it/s]

Exception in thread Thread-28:
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.10/dist-packages/ragas/executor.py", line 95, in run
    results = self.loop.run_until_complete(self._aresults())
  File "/usr/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/usr/local/lib/python3.10/dist-packages/ragas/executor.py", line 83, in _aresults
    raise e
  File "/usr/local/lib/python3.10/dist-packages/ragas/executor.py", line 78, in _aresults
    r = await future
  File "/usr/lib/python3.10/asyncio/tasks.py", line 571, in _wait_for_one
    return f.result()  # May raise f.exception().
  File "/usr/local/lib/python3.10/dist-packages/ragas/executor.py", line 37, in sema_coro
    return await coro
  File "/usr/local/lib/python3.10/dist-packages/ragas/executor.py", line 111, in wrapped_callable_async
    return counter, await callable(

ExceptionInRunner: The runner thread which was running the jobs raised an exeception. Read the traceback above to debug it. You can also pass `raise_exceptions=False` incase you want to show only a warning message instead.

In [ ]:
second_new_advanced_retrieval_results

In [ ]:
df_first = pd.DataFrame(list(new_advanced_retrieval_results.items()), columns=['Metric', 'GPT3.5 + TE3 + MQR'])
df_second = pd.DataFrame(list(second_new_advanced_retrieval_results.items()), columns=['Metric', 'GPT4o + TE3 + MQR'])

df_merged = pd.merge(df_first, df_second, on='Metric')

df_merged['Delta'] = df_merged['GPT4o + TE3 + MQR'] - df_merged['GPT3.5 + TE3 + MQR']

df_merged